hazelcast/src/main/java/com/hazelcast/datalink/DataLinkService.java

```diff
@@ -20,34 +20,89 @@
 import com.hazelcast.core.HazelcastException;
 import com.hazelcast.spi.annotation.Beta;
 
+import java.util.Map;
+
 /**
- * Service for accessing data link factories
+ * Service for managing {@link DataLink}s and their lifecycle.
+ * <p>
+ * {@link DataLink}s defined in the configuration are created at startup.
+ * {@link DataLink}s added to the configuration dynamically are created via
+ * {@link #createConfigDataLink(DataLinkConfig)}.
+ * {@link DataLink}s created via SQL are added via {@link #createSqlDataLink(String, String, Map)},
+ * these can be removed via {@link #removeDataLink(String)}.
+ * <p>
+ * When a new config is added via {@link #createConfigDataLink(DataLinkConfig)} is created and SQL DataLink already
+ * exists the existing DataLink is replaced and closed.
  *
- * @since 5.2
+ * @since 5.3
  */
 @Beta
 public interface DataLinkService extends AutoCloseable {
 
     /**
-     * Tests data link configuration.
+     * Creates a new DataLink with the given config
+     * <p>
+     * Such DataLink is considered immutable and can't be updated.
+     * The DataLink is closed when {@link #close()} is called.
+     *
+     * @param config the configuration of the DataLink
+     */
+    void createConfigDataLink(DataLinkConfig config);
+
+    /**
+     * Creates a new DataLink with the given parameters.
      *
-     * @param config name of the data link factory
-     * @return {@code true} if test was successful
-     * @throws Exception if the test operation fails
-     * @since 5.3
+     * @param name    name of the DataLink
+     * @param type    type of the DataLink
+     * @param options options configuring the DataLink
      */
-    boolean testConnection(DataLinkConfig config) throws Exception;
+    void createSqlDataLink(String name, String type, Map<String, String> options);
```

Comment:
The datalink is an abstract concept, so why do we have SQL-specific methods on the DatalinkService? What is then the point of having such an abstraction?

Reply:
As a user of that method, I may say - it's extremely convenient. 
The syntax of that command is :
```
CREATE [OR REPLACE] DATA LINK $name TYPE $type OPTIONS ('$opt_key' = '$opt_value')
```
As you can see, it's perfectly fits.

I wonder - is `DataLinkService` a public API?

Reply:
Since it is not behind an impl/internal package, it is API.

Reply:
We have two different classes of data links, depending on how you create them and they have different lifecycles.

One type is defined in the configuration (either static in file, or dynamic via `Config#addDataLink`/reload config etc..). For now, this datalink is immutable.

Then we have data links created via SQL interface - see `DataLinkService#createSqlDataLink(String, String, Map)`.

I felt distinguishing those two by using a different name is better than just based on the method parameters. I am open to some other name, maybe `createImmutableDataLink` and `createMutableDataLink` for the SQL one to better describe the difference between them (and not name them based on where they are used)?

wdyt?

Reply:
If `DataLinkService` is an internal API, I am fine with the current state.

Reply:
Reduced the API to getDataLink only (the rest moved to `InternalDataLinkService`
https://github.com/hazelcast/hazelcast/commit/a4f8d95bdc3b4325679daf468d3e695d8a8272db
https://github.com/hazelcast/hazelcast/commit/dbc14ce5f1f3db4f1276a7e78642a266a0b1f1d7